### PR TITLE
chore(internal): make some internal modules play nicer with tests that create threads

### PR DESCRIPTION
**Pull request by emmettbutler**
This pull request contains changes pulled out of https://github.com/DataDog/dd-trace-py/pull/4863. They're thematically unified in that they make little adjustments to the debugger and profiler to make them more amenable to the kind of testing that #4863 does. More detailed motivation for each change can be found in the commit messages.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Review by Yun-Kim — DISMISSED** [review]

**Review by Yun-Kim — APPROVED** [review]

File: ddtrace/profiling/collector/stack.pyx [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from __future__ import absolute_import
 
 import sys
-import threading
+import threading as ddtrace_threading
```
Comment (P403n1x87) [inline ddtrace/profiling/collector/stack.pyx:5]: The changes in this module are probably harder to justify without the full context of the referenced PR. So I'm thinking we can leave this source out?

**Comment by emmettbutler** [comment]
Here's the [specific test failure](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/28600/workflows/1e7f8ee1-cf90-4566-a419-e7a32096f2c4/jobs/1946231) that prompted the [removal](https://github.com/DataDog/dd-trace-py/pull/5102/commits/b3959b614a5ee7facc55d928b52d03b30170fc6a) of the no-warnings test

File: ddtrace/internal/compat.py [hunk 2]
```diff
@@ -86,26 +87,17 @@
 else:
     pattern_type = re._pattern_type  # type: ignore[misc,attr-defined]
 
-try:
-    from inspect import getargspec as getfullargspec
-
-    def is_not_void_function(f, argspec):
-        return argspec.args or argspec.varargs or argspec.keywords or argspec.defaults or isgeneratorfunction(f)
-
```
Comment (Yun-Kim) [inline ddtrace/internal/compat.py:102]: Can we keep `inspect.getargspec()` for Python 2?
```suggestion
if PY2:
    from inspect import getargspec as getfullargspec
    def is_not_void_function(f, argspec):
        return argspec.args or argspec.varargs or argspec.keywords or argspec.defaults or isgeneratorfunction(f)
else:
    from inspect import getfullargspec  # type: ignore[assignment]  # noqa: F401

    def is_not_void_function(f, argspec):
        return (
            argspec.args
            or argspec.varargs
            or argspec.varkw
            or argspec.defaults
            or argspec.kwonlyargs
            or argspec.kwonlydefaults
            or isgeneratorfunction(f)
        )
```

Comment (Yun-Kim) [reply ddtrace/internal/compat.py:102]: NVM - your change is fine as is.

File: ddtrace/profiling/profiler.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def _restart_on_fork(self):
         # Be sure to stop the parent first, since it might have to e.g. unpatch functions
         # Do not flush data as we don't want to have multiple copies of the parent profile exported.
         try:
-            self._profiler.stop(flush=False)
+            self._profiler.stop(flush=False, join=False)
```
Comment (brettlangdon) [inline ddtrace/profiling/profiler.py:87]: why do we not join here?

**Review by brettlangdon — COMMENTED** [review]
one question (maybe request for a code comment?), but otherwise lgtm

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/profiling/profiler.py (hunk 3 above)
Comment (emmettbutler) [reply ddtrace/profiling/profiler.py:87]: It's an instance where making the tests pass required maintaining the old behavior.

Comment (Yun-Kim) [reply ddtrace/profiling/profiler.py:87]: Previous behavior did not join (see `_ProfilerInstance._stop_service()`) by default, but in this PR it does now, so we're keeping the same behavior as previously.

**Pull request merged**